pytket/phir/sharding/sharder.py

```diff
@@ -144,28 +180,35 @@ def _build_shard(self, command: Command) -> None:
             # by looking at overlap of bits_written
             elif not shard.bits_written.isdisjoint(bits_written):
                 logger.debug("...adding shard dep %s -> WAW", shard.ID)
-                depends_upon.add(shard.ID)
+                add_dependency = True
 
             # Check for read-after-write (value seen would change if reordered)
             elif not shard.bits_written.isdisjoint(bits_read):
                 logger.debug("...adding shard dep %s -> RAW", shard.ID)
-                depends_upon.add(shard.ID)
+                add_dependency = True
 
             # Check for write-after-read (no reordering or read is changed)
-            elif not shard.bits_written.isdisjoint(bits_read):
+            elif not shard.bits_read.isdisjoint(bits_written):
                 logger.debug("...adding shard dep %s -> WAR", shard.ID)
+                add_dependency = True
+
+            if add_dependency:
                 depends_upon.add(shard.ID)
+                # Avoid recalculating for anything previous
+                shards_to_skip.update(self._get_dependencies(shard))
 
-        shard = Shard(
-            command,
-            sub_commands,
-            qubits_used,
-            bits_written,
-            bits_read,
-            depends_upon,
-        )
-        self._shards.append(shard)
-        logger.debug("Appended shard: %s", shard)
+        return depends_upon
+
+    def _get_dependencies(self, shard: Shard) -> set[int]:
+        dependencies: set[int] = set()
+        self._get_dependencies_recursive(dependencies, shard)
+        return dependencies
+
+    def _get_dependencies_recursive(self, accumulator: set[int], shard: Shard) -> None:
```

Comment:
Sorry, my comment above may have been confusing, currently this PR incurs significant cost and is not ready to be merged. `make tests` is taking way longer than usual (perhaps because of the issue with recursion that you just identified).

Compare the build times for the two recent PRs:
<img width="972" alt="Screenshot 2023-11-10 at 4 08 03 PM" src="https://github.com/CQCL/pytket-phir/assets/130700862/9addf2eb-8ee9-4488-a455-c20219251750">

Reply:
Not 100% sure but since this is clearly not a slam dunk I'm going to close it and try a more significant refactor that should reduce to linear time.